fix: fix copybook hover for nested structure #2886

amanPras opened the pull request:
## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test that #2886 is fixed
- [x] Test 2 #2857 continues to work
Open a cobol file with variables defined in a copybook and used in the main document
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. TESTA.
       ENVIRONMENT DIVISION.
       CONFIGURATION SECTION.
       SPECIAL-NAMES.
       INPUT-OUTPUT SECTION.
       FILE-CONTROL.
       DATA DIVISION.
        FILE SECTION.
        WORKING-STORAGE SECTION.
        COPY CPY1.
        LINKAGE SECTION.
       PROCEDURE DIVISION.
           display A1.
           STOP RUN.

``` 
CPY1.cpy
```cobol
        01 A.
           05 A1 PIC X.
``` 

- Go to line 14, Perform Go To Definition it will navigate to copybook 

- Perform Go To References

- Verify the result
## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

amanPras committed:
fix: fix copybook hover for nested structure #2886

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: run prettier

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -784,6 +784,57 @@ _NOTE: other versions exist due to replac(e/ing) or multiple use of same copyboo
+
+  test("hover over copybook - in a nested structure", async () => {
+    await helper.updateConfig("basic.json");
+    const editor = await helper.showDocument(
+      path.join("HOVER_COPYBOOK_NESTED.CBL"),
+    );
```
?
```suggestion
    const editor = await helper.showDocument("HOVER_COPYBOOK_NESTED.CBL");
```

slavek-kucera commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -784,6 +784,57 @@ _NOTE: other versions exist due to replac(e/ing) or multiple use of same copyboo
+    let editor_copybook: vscode.TextEditor | undefined;
+    await helper.waitFor(() => {
+      editor_copybook = vscode.window.activeTextEditor;
+      return !!editor_copybook;
```
Should we also make sure that `editor_copybook !== editor`?
```suggestion
      return !!editor_copybook && editor_copybook !== editor;
```

amanPras committed:
fix: fix references

added implementation looks in efficient, but the current node structure doesn't place the child nodes under CopyNode,
and we can't safely return from the first match

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/RangeUtils.java:
```diff
@@ -39,20 +39,31 @@ public static Optional<Node> findNodeByPosition(Node node, String uri, Position
       Optional<Node> nodeByPosition = findNodeByPosition(child, uri, position);
       if (nodeByPosition.isPresent()) {
-        return nodeByPosition;
+        candidate = nodeByPosition.get();
```
added implementation looks in efficient, but the current node structure doesn't place the child nodes under CopyNode,
and we can't safely guarantee that found node is most appropriately matched until we scan the entire structure.

amanPras commented on clients/cobol-lsp-vscode-extension/src/test/suite/lsp.spec.test.ts:
```diff
@@ -776,14 +776,100 @@ _NOTE: other versions exist due to replac(e/ing) or multiple use of same copyboo
       normalizeLineEndings(`\`\`\`cobol
-       PARAG2.
-           DISPLAY 'PARAG2'.
+       PARAG3.
```
Updated matches the last most valid position vs prev impl which matched the first most valid position.

amanPras committed:
chore: run prettier

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: fix hover text for section

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

# Conflicts:
#	server/engine/src/main/java/org/eclipse/lsp/cobol/core/visitor/CobolVisitor.java

amanPras commented:
Drafting due to performance degradation

amanPras committed:
fix: fix hover text for section and references - take 2

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: format code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/RangeUtils.java:
```diff
@@ -53,6 +55,54 @@ public static Optional<Node> findNodeByPosition(Node node, String uri, Position
+   * @param position a cursor position
+   * @return the found node
+   */
+  public static List<Node> findAllApplicableNodesByPosition(
```
This may be used to show all hover informations as well (in case of multiple copyb statements or replacing), but didn't want to complicate this PR

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/RangeUtils.java:
```diff
@@ -53,6 +55,54 @@ public static Optional<Node> findNodeByPosition(Node node, String uri, Position
+      }
+      List<Node> nodesByPosition = findAllApplicableNodesByPosition(child, uri, position);
+      if (!nodesByPosition.isEmpty()) {
+        result.addAll(nodesByPosition);
```
You can avoid these temporary result lists by
```java
public static List<Node> findAllApplicableNodesByPosition(Node node, String uri, Position position) {
    List<Node> result = new ArrayList<>();
    findAllApplicableNodesByPositionImpl(result, node, uri, position);
    return result;
}

private static void findAllApplicableNodesByPositionImpl(List<Node> result, Node node, String uri, Position position) {
    Node candidate = null;
    for (Node child : node.getChildren()) {
      if (isInside(uri, position, child.getLocality())) {
        candidate = getCandidate(uri, child, candidate);
      }
      findAllApplicableNodesByPositionImpl(result, child, uri, position);
    }
    if (candidate != null) {
      result.add(candidate);
    }
}
```

amanPras replied:
updated

amanPras committed:
fix: apply changes as suggested in PR

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: format code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: show all hovers for a position linked to multiple nodes

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: format code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: fix test

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
fix: fix references for variable definitions

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: format code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: format code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

amanPras committed:
chore: format code

Signed-off-by: amanprashant <aman.prashant@broadcom.com>

slavek-kucera commented on server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/RangeUtils.java:
```diff
@@ -39,20 +42,77 @@ public static Optional<Node> findNodeByPosition(Node node, String uri, Position
+    }
+    if (candidate != null) {
+      Node finalCandidate = candidate;
+      boolean shouldNotAdd =
+          result.stream()
+              .map(n -> n.getNearestParent(finalCandidate::equals))
+              .anyMatch(Optional::isPresent);
+      if (!shouldNotAdd && finalCandidate instanceof DefinedAndUsedStructure) {
```
```suggestion
      boolean shouldAdd =
          result.stream()
              .map(n -> n.getNearestParent(finalCandidate::equals))
              .noneMatch(Optional::isPresent);
      if (shouldAdd && finalCandidate instanceof DefinedAndUsedStructure) {
```

amanPras committed:
Update server/common/src/main/java/org/eclipse/lsp/cobol/common/utils/RangeUtils.java

Co-authored-by: slavek-kucera <53339291+slavek-kucera@users.noreply.github.com>
